feat: flatten *List objects in Helm/Kustomize

schradert opened the pull request:
These are not actual Kubernetes resource definitions and are not validated anyways. They also break evaluation because they do not require a metadata.name property.

schradert commented:
List objects from upstream Helm charts were failing [here](https://github.com/arnarg/nixidy/blob/7d8acc7b4418f2b976f9f948b4e05b87f538bfe7/modules/build.nix#L12) because they lack `metadata.name`. I didn't think up a good heuristic for file naming in case there are multiple in an application, but I also thought this might be an interesting albeit unexpected approach to validating the contents of the List.

arnarg commented on modules/applications/helm.nix:
```diff
@@ -94,6 +94,12 @@ in {
   config = with lib; let
+    # Recursively flatten *List objects
+    flattenListObjects = builtins.concatMap (object:
+      if builtins.match "^.*List$" object.kind != null
```
I think this if statement should also check either if the object has `items` (which is a list) or the lack of `metadata`, or both. Otherwise some arbitrary CRD with a kind that ends with List would also be caught here.

arnarg reviewed (changes requested):
Thanks for this change! Looks useful! And thanks for writing tests for it :)

Since the function `flattenListObjects` is identical between the two files, could you add it to helpers like I did with `getGVK`?

schradert commented:
Okay I refactored and fixed the filter. However, the new test I added is failing with kustomize, and I'm not sure why. `expression` evaluates to `[]` instead of the value of `expected` written

arnarg reviewed (changes requested):
So... I had some time to look into this and it seems that a simple update of nixpkgs input fixes the issue where `CustomList` isn't being included in the build output. I also noticed that kustomize will flatten the `ConfigMapList` for us (maybe the older version of kubectl was handling `CustomList` as a `*List` object and not finding `.items`).

So I updated all the flake inputs on main and I therefore suggest you rebase on top of that and maybe only include `flattenListObjects` in `modules/applications/helm.nix` as it's unnecessary in kustomize.
